tooling/bash-limits.md

```diff
@@ -0,0 +1,228 @@
+# BASH Limits
+What are good indicators your BASH script has become too complex and you should stop to re-implement your system in a "real" language?
+
+### Index
+1. ["Real" Languages](#real-languages)
+    1. [Examples](#examples)
+1. [Red Flags](#red-flags)
+1. [Process](#process)
+
+## "Real" Languages
+BASH is a real language, a Turing-complete, interpreted scripting language. This is why "real" is in quotes throughout this document.
+
+However, it's important to note that while BASH is Turing-complete, it is not a general-purpose programming language and it's not designed to be used for large, complex projects or for tasks that require advanced features like object-oriented programming, closures, currying, etc.
+
+> Throughout this document, when I have referred to "real" languages, I have been using it to refer to programming languages that are typically used for general-purpose programming and have a large number of features and capabilities. These languages are generally more versatile and expressive than languages like BASH which are primarily used for system administration and command-line scripting.
+>
+> These languages are designed to be used for a wide range of tasks, from web development and data processing to machine learning and artificial intelligence. They are also more suited for large, complex projects, and have more robust tools for debugging, testing, and deploying code. They often provide support for object-oriented, functional and other programming paradigms, and have a larger user base and libraries.
+>
+> In contrast, BASH is a command-line scripting language that is primarily used for system administration and automating tasks. It's a powerful tool for working with the command-line and performing tasks like text processing, file management, and process management. But when it comes to more advanced features like object-oriented programming, closures, currying, etc. it lacks the capabilities of "real" languages.
+>
+> It's important to note that BASH can be used in conjunction with other languages, like Python, to accomplish more complex tasks and to be more expressive, but when the task becomes too complex or the codebase becomes too big, it's better to use a more powerful language that is better suited for the task.
+
+### Examples
+Examples of "real" languages.
+- C#
+- C++
+- Go
+- Haskell
+- Java
+- JavaScript
+- Python
+- Rust
+
+Which languages specifically are acceptable in our repos or organization is a wholly different discussion for a different document.
+
+## Red Flags
+Here are some suggestions.
+1. **Too Many Lines of Code**.
+1. **Complex mathematical calculations**, any math involving decimals, or any math where a third-party program is invoked because BASH cannot do it.
```

Comment:
Please see [Unix Philosophy](https://en.wikipedia.org/wiki/Unix_philosophy#Origin). bash is a fine place to call other programs, such as `bc`, so I think rewording this to something like `implementing complicated floating point mathematical algorithms wholly within BASH.` would get your actually point across without losing credibility for denying the Unix Philosophy.

Reply:
As I shared on IM, here is an objectively false response returned by `bc`.
```bash
$ echo 23/8 | bc
2
```
Compare this to a "real" language.

![2023-01-18 16-12-40 - basic math in a real language](https://user-images.githubusercontent.com/34947245/216157963-c1edc435-8ed4-436c-99af-3191e3b6dd55.png)

You suggested the following code snippet.

```bash
echo "scale=4; 23/8" | bc
```

While this does return a correct solution to the math problem, _it requires one to know the correct solution in order to write this code_. Consider a script parsing metrics, for example, where the input can be arbitrary. This is extremely error-prone, and is not an appropriate level of risk for a production-facing applications.

These were my responses when we had this discussion the first time via IM on 2023-01-18.

How come this feedback was not raised on 2023-01-24 when the whole Engineering team ratified this document? You were on that video call.

Reply:
I'd already raised the issue and my energy for it at that time was low. You don't need my review, feel free to disregard.

As long as this phrase exist in the document, I can not approve it: 
`[bash should not be used if] a third-party program is invoked because BASH cannot do it.`

Are you suggesting bash shouldn't call grep?

Reply:
23/8 in a "real" language:
```
beauty$ cat go.cpp 
#include <iostream>

int main() {
   int x = 23.0;
   int y = 8.0;
   float f = x/y;
   std::cout << f << '\n';
}
beauty$ g++     go.cpp   -o go
beauty$ ./go 
2
```
Probably should have turned on some warnings. :-) 

But my point is this: if you misuse a tool, you might end up with bad results.

Reply:
> `[bash should not be used if] a third-party program is invoked because BASH cannot do it.`
> Are you suggesting bash shouldn't call grep?

This is a misrepresentation of what that line said.

Regardless, I have updated the document to narrow the scope of this list element somewhat and to make it more specific. I have also added a note explaining _why_ BASH is undesirable for complex math operations. Please review these changes and let me know what you think.

Reply:
This looks much better.